### PR TITLE
Make Measurements Pytrees

### PR DESCRIPTION
**Pull request by albi3ro**
This PR registers all `MeasurementProcess` objects as jax pytrees. [sc-40588]

### CODE REVIEW

**Review by mudit2812 — APPROVED** [review]

**Review by vincentmr — APPROVED** [review]
Looks good to me, cheers @albi3ro .

**Review by timmysilv — APPROVED** [review]
looks awesome, also helped highlight some little things to touch up!

quick question: (un/)flattening should help with copying and replacing operators, right? Curious because we [replace measurements in-place](https://github.com/PennyLaneAI/pennylane/blob/c60d8bca58a4cff4168fea61b38a64a74d6e7b88/pennylane/devices/qubit/preprocess.py#L132-L137) in device preprocessing, and I'm wondering if that code can benefit from this change

**Comment by albi3ro** [comment]
> Not necessarily in the scope of the PR, but I was wondering whether we would like to leverage `_flatten` in dyadics like `==` or `_equal`?

Potentially we could rewrite a lot of `qml.equal` (used by `__eq__`). Tensor data has to use `qml.math.allclose` instead of `==`, so it wouldn't necessarily be as easy as `type(obj1) == type(obj2) and obj1._flatten() == obj2._flatten()`, but it might still be a better solution than what we currently have.

File: pennylane/measurements/__init__.py [hunk 1]
```diff
@@ -157,6 +157,23 @@ def circuit(x):
     functions will return an instance of the corresponding measurement process
     (e.g. :class:`CountsMP`). This decision is just for design purposes.
 
+.. details::
+    :title: Serialization and Pytree format
+    :href: serialization
+
+    PennyLane measurements are automatically registered as `Pytrees <https://jax.readthedocs.io/en/latest/pytrees.html>`_ .
+
+    The :class:`~.MeasurementProcess` definitions are sufficient for all PL measurements.
```
Comment (trbromley) [inline pennylane/measurements/__init__.py:166]: So does this mean that users creating custom measurements do not need to do anything extra?

Comment (albi3ro) [reply pennylane/measurements/__init__.py:166]: Oops. Actually that comment is wrong. They need to be overridden if the measurement process has extra metadata.

Comment (albi3ro) [reply pennylane/measurements/__init__.py:166]: I'll fix that in the tape pytree PR.

**Pull request merged**